### PR TITLE
iox-#605 Refactor `PointerRepository`

### LINKED ISSUE #605: Refactor Relocatable Pointer and Relative Pointer
## Brief feature description

The relocatable pointer (`iceoryx_utils/internal/relocatable_pointer`) implementation needs to be cleaned up:

 1. [x] add noexcept, const, replace `size_t` with `uint64_t` (done in #618 except making parameters const)
 2. [x] separate implementation and declaration for all classes (#618)
 3. [x] one class per header file (#618)
 4. [x] add subnamespace `iox::rp` where everything is stored (#618)
 5. [x] ~~restructure the unit tests / add missing tests~~ Moved to #1745
 6. [x] add doxygen documentation (#618)
 7. [x] remove includes of base_relative_ptr.hpp & base_relocatable_ptr.hpp
 9. [x] ~~consider removing ``ptr_t`` and ``const_ptr_t`` aliases and use ``void*`` directly (see https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597604809)~~
 10. [x] ~~add ``void* get() noexcept``, ``const void* get() const noexcept``, ``void* getBasePtr() noexcept`` and ``const void* getBasePtr() const noexcept`` to BaseRelativePointer~~ -> Not implemented as `SharedPointer` uses `get() const noexcept`
 11. [x] ~~consider changing return type of ``PointerRepository::unregisterPrt()`` to ``void`` (see https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597609248)~~ - Note: maybe not, it might be useful to have an indication whether the deregistration was successful
 12. [x] move duplicated code into a separate method (see https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597613171)
 13. [x] rename ``PointerRepository::registerPtr()`` methods if both are needed (see https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597613855) + consider returning a ``cxx::unique_ptr<id_t>``
 14. [x] check invalid id logic (see https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597617681 and https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597619467)
 15. [x] ~~add const methods to RelativePointer (see https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597620985, https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597621144 and https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597621263).~~ This might not be needed, compare with e.g. std::shared_pointer https://en.cppreference.com/w/cpp/memory/shared_ptr/operator* It should work similar to STL smart pointers.
 16. [x] invalidate ``other`` move c'tor and move assignment operator of BaseRelocatablePointer (see https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597626715 and https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597628041)
 17. [x] use composition instead of inheritance (see https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r599596822 and https://github.com/eclipse-iceoryx/iceoryx/issues/605#issuecomment-806045008) - if done, consider renaming Base...Pointer to Untyped...Pointer
 18. [x] ~~replace `RelativePointer` by `RelocatablePointer` where applicable, e.g. `Semaphore` and `LoFFLi`~~ Moved to #1745
 19. [x] Move `RelativePointer` to public include

### PR DESCRIPTION
**Pull request by mossmaurice**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer
* Remove `INVALID_ID` and introduce `RAW_POINTER_BEHAVIOUR`
* Rename `registerPtr` method and return `cxx::optional`
* Remove `isValid` and `print` methods
* Add sanity check if user didn't call `init` when using `LoFFLi`

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #605 (partly)

### CODE REVIEW

File: iceoryx_posh/include/iceoryx_posh/internal/mepoo/mepoo_segment.inl [hunk 1]
```diff
@@ -76,8 +76,12 @@ inline SharedMemoryObjectType MePooSegment<SharedMemoryObjectType, MemoryManager
...
             .create()
             .and_then([this](auto& sharedMemoryObject) {
-                this->setSegmentId(static_cast<uint64_t>(iox::rp::BaseRelativePointer::registerPtr(
-                    sharedMemoryObject.getBaseAddress(), sharedMemoryObject.getSizeInBytes())));
+                auto maybeSegmentId = iox::rp::BaseRelativePointer::registerPtr(sharedMemoryObject.getBaseAddress(),
+                                                                                sharedMemoryObject.getSizeInBytes());
+                if (maybeSegmentId.has_value())
+                {
+                    this->setSegmentId(static_cast<uint64_t>(maybeSegmentId.value()));
+                }
```
Comment (mossmaurice) [inline iceoryx_posh/include/iceoryx_posh/internal/mepoo/mepoo_segment.inl:84]: @MatthiasKillat As the sad path is a fatal error, I would create a new error enum entry here. What do you think?

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp [hunk 2]
```diff
@@ -64,11 +64,13 @@ class RelativePointer : public BaseRelativePointer
     template <typename U = T>
     typename std::enable_if<!std::is_void<U>::value, const U&>::type operator*() const noexcept;
 
-    /// @brief read-only access to the underlying object
+    /// @brief read-only access to the underlying object. If the RelativePointer does not point to anything the
+    /// application might terminate.
     /// @return a pointer to the underlying object
     T* operator->() const noexcept;
```
Comment (elfenpiff) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp:70]: This breaks the const correctness and it should be 
`const T* operator->() const noexcept;`
Furthermore, we should introduce a `T* operator->() noexcept` to provide write access.

I know that the STL `unique_ptr` and `shared_ptr` does it differently but here I totally disagree with their behavior. I know that the relative pointer does not own the memory but it provide access to it. When a function would like to have a pointer as argument one can make sure with:
```cxx
void myFunc(const int * ptr);
```
That it is not changed but in here it would mean to generate a completely new type `RelativePointer<const T>` where the memory is always const. But most likely users will misuse it quickly and assume:
```cxx
void myFunc(const RelativePointer<int> ptr);
```
would assume that `myFunc` cannot change the content when it actually can.

See also the `cxx::unique_ptr` where we already ensure the const correctness with the proposal from above.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp [hunk 3]
```diff
@@ -64,11 +64,13 @@ class RelativePointer : public BaseRelativePointer
...
 
-    /// @brief access the underlying object
+    /// @brief access the underlying object. If the RelativePointer does not point to anything the
+    /// application might terminate.
     /// @return a pointer to the underlying object
     T* get() const noexcept;
```
Comment (elfenpiff) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp:75]: Same here, breaks the const correctness and it should be:
```cxx
const T* get() const noexcept;
T* get() noexcept;
```

**Review by elfenpiff — CHANGES_REQUESTED** [review]
I would propose to add the suggested methods from my comment in this pull request. This should only add ~20 lines of additional code and 2 more unit tests for the const case.

As an alternative I would also be happy with an issue where this is tracked but I think writing the issue, tracking and following up on it takes more time than actually fixing it right away.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp (hunk 2 above)
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp:70]: @elfenpiff I know what you mean, but I still disagree here. I believe a lot of effort has gone into designing the STL APIs and it's also quite proven in use. We might not even be aware for certain corner cases.

I would suggest to leave it as it is and also adapt it accordingly in `cxx::unique_ptr` as the user can control the constness like:

```
RelativePointer<T> p             ===> T* p
const RelativePointer<T> p       ===> T* const p
RelativePointer<const T> p       ===> const T* p
const RelativePointer<const T> p ===> const T* const p
```

To prevent the potential misuse, I would suggest to add the above to doxygen. What do you think?

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp (hunk 3 above)
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp:75]: Closing this thread as it's a duplicate of [this one](https://github.com/eclipse-iceoryx/iceoryx/pull/1599#discussion_r963478931).

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp (hunk 2 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp:70]: @mossmaurice This is the intention and consistent with other fancy pointer APIs to my knowledge, so we should keep it like this. Adding the comment would help.

@elfenpiff The user assumption would be wrong then and she should use 

```
void myFunc(RelativePointer<const int> ptr);
```
The system is consistent and documentation is available.

Another change I would make is revert the name to its original `relative_ptr` for STL style consistency if possible given the current tooling (was like this before some changes before this PR).

File: doc/website/release-notes/iceoryx-unreleased.md [hunk 4]
```diff
@@ -362,6 +363,25 @@
...
     constexpr uint32_t MAX_VAL = algorithm::maxVal(3, 1890, 57);
     constexpr uint32_t MIN_VAL = algorithm::minVal(3, 1890, 57);
+
+20. `ReleativePointer::registerPtr` returns `cxx::optional`
+
+    ```cpp
+    // before
+    uint64_t id = RelativePointer::register(myPtr, 42);
+
+    if(id == INVALID_ID)
```
Comment (MatthiasKillat) [inline doc/website/release-notes/iceoryx-unreleased.md:373]: Note that normally that should never happen if there is sufficient space in the repository. If we see that the current setting is too limiting we need to increase it, but it is optimized for thousands of ids but not millions (brute force search during registration, but we could hash later if needed).

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/pointer_repository.hpp [hunk 5]
```diff
@@ -45,33 +46,40 @@ class PointerRepository
...
 
     static_assert(MAX_ID >= MIN_ID, "MAX_ID must be greater or equal than MIN_ID!");
     static_assert(CAPACITY >= 2, "CAPACITY must be at least 2!");
 
   public:
-    static constexpr id_t INVALID_ID = std::numeric_limits<id_t>::max();
+    /// @note 0 is a special purpose id and reserved
+    /// id 0 is reserved to interpret the offset just as a raw pointer,
+    /// i.e. its corresponding base ptr is 0
+    static constexpr id_t RAW_POINTER_BEHAVIOUR{0};
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/pointer_repository.hpp:59]:
```suggestion
    static constexpr id_t RAW_POINTER_BEHAVIOUR_ID{0};
```

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/pointer_repository.hpp [hunk 6]
```diff
@@ -45,33 +46,40 @@ class PointerRepository
...
+    ~PointerRepository() noexcept = default;
+
+    PointerRepository(const PointerRepository&) = delete;
+    PointerRepository(PointerRepository&&) = delete;
+    PointerRepository& operator=(const PointerRepository&) = delete;
+    PointerRepository& operator=(PointerRepository&&) = delete;
 
     /// @brief registers the start pointer of the segment in another application with a specific id
-    /// @param[in] id identifies the segment
+    /// @param[in] id identifies the segment which the pointer should be added to
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/pointer_repository.hpp:71]:
```suggestion
    /// @param[in] id identifies the segment that the pointer should be added to
```

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/pointer_repository.hpp [hunk 7]
```diff
@@ -45,33 +46,40 @@ class PointerRepository
...
     /// @param[in] ptr is the start pointer of the segment
     /// @param[in] size is the size of the segment
     /// @return true if the registration was successful, otherwise false
-    bool registerPtr(id_t id, ptr_t ptr, uint64_t size) noexcept;
+    bool registerPtrWithId(id_t id, ptr_t ptr, uint64_t size) noexcept;
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/pointer_repository.hpp:75]: I prefer overloading to long names. This is how it was done in C in the past (and had to, because overloading did not exist).

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/pointer_repository.hpp [hunk 8]
```diff
@@ -45,33 +46,40 @@ class PointerRepository
...
 
     /// @brief registers the start pointer of a segment with a specific size
     /// @param[in] ptr is the start pointer of the segment
     /// @param[in] size is the size of the segment
-    /// @return the id that identifies the segment
-    id_t registerPtr(const ptr_t ptr, uint64_t size = 0U) noexcept;
+    /// @return the segment id wrapped in an cxx::optional to which the pointer was added, cxx::nullopt if pointer was
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/pointer_repository.hpp:80]:
```suggestion
    /// @return the segment id to which the pointer was added wrapped in an cxx::optional, cxx::nullopt if pointer was
```

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/pointer_repository.inl [hunk 9]
```diff
@@ -71,11 +72,11 @@ inline id_t PointerRepository<id_t, ptr_t, CAPACITY>::registerPtr(const ptr_t pt
             {
                 m_maxRegistered = id;
             }
-            return id;
+            return cxx::make_optional<id_t>(id);
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/pointer_repository.inl:75]: I think it is not needed, we can just use 
```suggestion
            return id;
```
and the ctor will be invoked (as the compiler knows the return type). Unless it is explicit ... which is another reason I do not like explicit as this verbosity does not help here. While reading the impl code the only relevant information is that some id is returned (and not a null-value). We only care about the wrapper type at call site.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/pointer_repository.inl [hunk 10]
```diff
@@ -132,30 +133,8 @@ inline id_t PointerRepository<id_t, ptr_t, CAPACITY>::searchId(ptr_t ptr) const
         }
     }
     /// @note implicitly interpret the pointer as a regular pointer if not found
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/pointer_repository.inl:135]:
```suggestion
    /// @note implicitly interpret the pointer as a regular pointer if not found
```
```suggestion
    /// @note treat the pointer as a regular pointer if not found
```

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp [hunk 11]
```diff
@@ -64,11 +64,13 @@ class RelativePointer : public BaseRelativePointer
...
 
-    /// @brief read-only access to the underlying object
+    /// @brief read-only access to the underlying object. If the RelativePointer does not point to anything the
+    /// application might terminate.
     /// @return a pointer to the underlying object
     T* operator->() const noexcept;
 
-    /// @brief access the underlying object
+    /// @brief access the underlying object. If the RelativePointer does not point to anything the
+    /// application might terminate.
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp:73]: I think id **does terminate** here currently for a logical nullptr, so we should state this. Note that I think it is not a good idea to do so, because just getting the pointer (as `nullptr`)) is not the problem. The problem is dereferencing it later, as with regular pointers so we should remove the check.

File: iceoryx_posh/source/roudi/memory/memory_provider.cpp [hunk 12]
```diff
@@ -86,14 +86,19 @@ cxx::expected<MemoryProviderError> MemoryProvider::create() noexcept
 
     m_memory = memoryResult.value();
     m_size = totalSize;
-    m_segmentId = rp::BaseRelativePointer::registerPtr(m_memory, m_size);
+    auto maybeSegmentId = rp::BaseRelativePointer::registerPtr(m_memory, m_size);
+
+    if (maybeSegmentId.has_value())
```
Comment (MatthiasKillat) [inline iceoryx_posh/source/roudi/memory/memory_provider.cpp:91]: If it does not work it can be considered fatal and we would actually not be able to continue. This cannot really happen for a reasonable number of segments though.

We should generate a fatal error in this case, or at least have a placeholder terminate and a todo for it (will add some `IOX_PANIC` or similar once available) .

File: iceoryx_posh/include/iceoryx_posh/internal/mepoo/mepoo_segment.inl (hunk 1 above)
Comment (MatthiasKillat) [reply iceoryx_posh/include/iceoryx_posh/internal/mepoo/mepoo_segment.inl:84]: @mossmaurice I would at least add a terminate, a todo and later call `IOX_PANIC` (for a fatal error). But we can also have an error code for it (INSUFFICIENT_SEGMENT_IDS or something like this).

File: iceoryx_hoofs/source/concurrent/loffli.cpp [hunk 13]
```diff
@@ -55,7 +55,7 @@ bool LoFFLi::pop(Index_t& index) noexcept
     do
     {
         // we are empty if next points to an element with index of Size
-        if (oldHead.indexToNextFreeIndex >= m_size)
+        if (oldHead.indexToNextFreeIndex >= m_size || !m_nextFreeIndex)
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/source/concurrent/loffli.cpp:58]: Why is this needed?

Note that the whole index terminology in loffli is kind of inconsistent or broken which makes it hard to follow (not well-defined, as there are at least two levels of indices from what I remember).

File: doc/website/release-notes/iceoryx-unreleased.md [hunk 14]
```diff
@@ -362,6 +363,25 @@
     #include "iceoryx_hoofs/cxx/algorithm.hpp"
     constexpr uint32_t MAX_VAL = algorithm::maxVal(3, 1890, 57);
     constexpr uint32_t MIN_VAL = algorithm::minVal(3, 1890, 57);
+
+20. `ReleativePointer::registerPtr` returns `cxx::optional`
+
+    ```cpp
+    // before
+    uint64_t id = RelativePointer::register(myPtr, 42);
```
Comment (MatthiasKillat) [inline doc/website/release-notes/iceoryx-unreleased.md:371]:
```suggestion
    uint64_t id = RelativePointer::register(startAddress, numBytes);
```

File: doc/website/release-notes/iceoryx-unreleased.md [hunk 15]
```diff
@@ -362,6 +363,25 @@
...
+
+    if(id == INVALID_ID)
+    {
+        // Early exit
+    }
+
+    // after
+    auto maybeId = RelativePointer::register(myPtr, 42);
```
Comment (MatthiasKillat) [inline doc/website/release-notes/iceoryx-unreleased.md:379]:
```suggestion
    auto maybeId = RelativePointer::register(startAddress, numBytes);
```

We register a segment beginning at `startAddress` with size `numBytes` and get some id now associated with this segment (if successful).

**Review by MatthiasKillat — CHANGES_REQUESTED** [review]

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/pointer_repository.hpp (hunk 7 above)
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/pointer_repository.hpp:75]: @MatthiasKillat In general I agree with you, but in this case the parameters are rather similar so it's nicer if you type `.` in your IDE and then get two different methods offered.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp (hunk 11 above)
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp:73]: @MatthiasKillat True, I moved the assert-style check to `operator->()` and adapted the wording.

File: iceoryx_hoofs/source/concurrent/loffli.cpp (hunk 13 above)
Comment (mossmaurice) [reply iceoryx_hoofs/source/concurrent/loffli.cpp:58]: @MatthiasKillat By change, I ran into an issue in `LoFFLi` when `init()` was not called. This check prevents `pop()` to work on the data if the `RelativePointer` was not initialized.

**Review by FerdinandSpitzschnueffler — COMMENTED** [review]
I wonder why you didn't need to adapt the relative ptr test files.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/pointer_repository.hpp [hunk 16]
```diff
@@ -45,33 +46,40 @@ class PointerRepository
...
+    PointerRepository& operator=(PointerRepository&&) = delete;
 
     /// @brief registers the start pointer of the segment in another application with a specific id
-    /// @param[in] id identifies the segment
+    /// @param[in] id identifies the segment that the pointer should be added to
     /// @param[in] ptr is the start pointer of the segment
     /// @param[in] size is the size of the segment
     /// @return true if the registration was successful, otherwise false
-    bool registerPtr(id_t id, ptr_t ptr, uint64_t size) noexcept;
+    bool registerPtrWithId(id_t id, ptr_t ptr, uint64_t size) noexcept;
```
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/pointer_repository.hpp:75]: They can. But value parameters should not be and I am fairly sure static code analysis agrees, unless this is disabled.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp [hunk 17]
```diff
@@ -64,11 +64,12 @@ class RelativePointer : public BaseRelativePointer
...
     typename std::enable_if<!std::is_void<U>::value, const U&>::type operator*() const noexcept;
 
-    /// @brief read-only access to the underlying object
+    /// @brief read-only access to the underlying object. If the RelativePointer does not point to anything the
+    /// application terminates.
     /// @return a pointer to the underlying object
     T* operator->() const noexcept;
 
-    /// @brief access the underlying object
+    /// @brief access the underlying object.
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp:72]: Actually 

```suggestion
    /// @brief Access the underlying object.
```

but this is inconsistent with the other documentation so we should not change it.

File: iceoryx_hoofs/source/concurrent/loffli.cpp (hunk 13 above)
Comment (mossmaurice) [reply iceoryx_hoofs/source/concurrent/loffli.cpp:58]: As there is already a todo, I didn't add a further one.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/base_relative_pointer.hpp [hunk 18]
```diff
@@ -124,7 +124,7 @@ class BaseRelativePointer
     /// @param[in] ptr starting address of the segment to be registered
     /// @param[in] size is the size of the segment
     /// @return id it was registered to
-    static id_underlying_t registerPtr(const ptr_t ptr, uint64_t size = 0U) noexcept;
+    static cxx::optional<id_underlying_t> registerPtr(const ptr_t ptr, uint64_t size = 0U) noexcept;
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/base_relative_pointer.hpp:127]: @FerdinandSpitzschnueffler As `BaseRelativePointer` will be merged with `RelativePointer` I would prefer to change doxygen on #1603

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/pointer_repository.hpp [hunk 19]
```diff
@@ -18,10 +18,11 @@
 #ifndef IOX_HOOFS_RELOCATABLE_POINTER_POINTER_REPOSITORY_HPP
 #define IOX_HOOFS_RELOCATABLE_POINTER_POINTER_REPOSITORY_HPP
 
+#include "iceoryx_hoofs/cxx/optional.hpp"
 #include "iceoryx_hoofs/cxx/vector.hpp"
-#include <iostream>
 
 #include <cassert>
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/pointer_repository.hpp:25]: @FerdinandSpitzschnueffler Good catch, thanks! Removed.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp (hunk 17 above)
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp:72]: @MatthiasKillat I capitalized all sentences in the beginning.

**Review by MatthiasKillat — DISMISSED** [review]

File: iceoryx_hoofs/source/concurrent/loffli.cpp (hunk 13 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/source/concurrent/loffli.cpp:58]: My inner chicken fears that we will forget to remove the unnecessary checks. Could you please add a hint to the todo in the header file? Something like "move init code to c'tor, remove unnecessary `!m_nextfreeIndex` check".

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by MatthiasKillat — APPROVED** [review]
